lib/services/role.go

```diff
@@ -1758,6 +1514,85 @@ func (set RoleSet) CheckLoginDuration(ttl time.Duration) ([]string, error) {
 	return out, nil
 }
 
+// CheckAccessToRemoteCluster checks if a role has access to remote cluster. Deny rules are
+// checked first then allow rules. Access to a cluster is determined by
+// namespaces, labels, and logins.
+//
+// Note, logging in this function only happens in debug mode, this is because
+// adding logging to this function (which is called on every server returned
+// by GetNodes) can slow down this function by 50x for large clusters!
+func (set RoleSet) CheckAccessToRemoteCluster(rc RemoteCluster) error {
+	var errs []error
+
+	rcLabels := rc.GetMetadata().Labels
+
+	// If there is no role in the set with labels and the cluster
+	// is not labelled ,assume that the role set has access to the cluster.
+	usesLabels := false
+	for _, role := range set {
+		if len(role.GetClusterLabels(Allow)) != 0 || len(role.GetClusterLabels(Deny)) != 0 {
+			usesLabels = true
+			break
+		}
+	}
+
+	if usesLabels == false && len(rcLabels) == 0 {
+		if log.GetLevel() == log.DebugLevel {
+			log.WithFields(log.Fields{
+				trace.Component: teleport.ComponentRBAC,
+			}).Debugf("Grant access to cluster %v - no role in %v uses cluster labels and the cluster is not labeled.",
+				rc.GetName(), set.RoleNames())
+		}
+		return nil
+	}
+
+	// Check deny rules first: a single matching namespace, label, or login from
+	// the deny role set prohibits access.
+	for _, role := range set {
+		matchLabels, labelsMessage, err := MatchLabels(role.GetClusterLabels(Deny), rcLabels)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if matchLabels == true {
+			if log.GetLevel() == log.DebugLevel {
+				log.WithFields(log.Fields{
+					trace.Component: teleport.ComponentRBAC,
+				}).Debugf("Access to cluster %v denied, deny rule in %v matched; match(label=%v)",
+					rc.GetName(), role.GetName(), labelsMessage)
+			}
+			return trace.AccessDenied("access to cluster denied")
+		}
+	}
+
+	// Check allow rules: namespace, label, and login have to all match in
+	// one role in the role set to be granted access.
+	for _, role := range set {
+		matchLabels, labelsMessage, err := MatchLabels(role.GetClusterLabels(Allow), rcLabels)
+		log.WithFields(log.Fields{
+			trace.Component: teleport.ComponentRBAC,
+		}).Debugf("Check access to role(%v) rc(%v, labels=%v) matchLabels=%v, msg=%v, err=%v allow=%v rcLabels=%v",
+			role.GetName(), rc.GetName(), rcLabels, matchLabels, labelsMessage, err, role.GetClusterLabels(Allow), rcLabels)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if matchLabels == true {
```

Comment:
```suggestion
		if matchLabels {
```

Reply:
@awly so we are no longer doing `matchLabels == true`?

Reply:
Nope, it was never preferred.
The only places where we keep that is for truth tables (exhaustive switches with multiple bools), because a few folks on the team prefer that.